src/irr/video/COpenGLPipelineCache.h

```diff
@@ -33,52 +33,75 @@ class COpenGLPipelineCache final : public IGPUPipelineCache
 
 	void merge(uint32_t _count, const IGPUPipelineCache** _srcCaches) override
 	{
-		const std::lock_guard<std::mutex> _(m_mutex);
+		const std::lock_guard<std::mutex> _1_(m_bin_cache_mutex);
+		const std::lock_guard<std::mutex> _2_(m_parsed_cache_mutex);
 
 		for (uint32_t i = 0u; i < _count; ++i)
 		{
-			const auto& src = static_cast<const COpenGLPipelineCache*>(_srcCaches[i])->m_cache;
-			m_cache.insert(src.begin(), src.end());
+			{
+				const auto& src = static_cast<const COpenGLPipelineCache*>(_srcCaches[i])->m_cache;
+				m_cache.insert(src.begin(), src.end());
+			}
+			{
+				const auto& src = static_cast<const COpenGLPipelineCache*>(_srcCaches[i])->m_parsedSpirvs;
+				m_parsedSpirvs.insert(src.begin(), src.end());
+			}
 		}
 	}
 
 	core::smart_refctd_ptr<asset::ICPUPipelineCache> convertToCPUCache() const override;
 
 	COpenGLSpecializedShader::SProgramBinary find(const SCacheKey& _key, const COpenGLPipelineLayout* _layout) const
 	{
-		const std::lock_guard<std::mutex> _(m_mutex);
+		const std::lock_guard<std::mutex> _(m_bin_cache_mutex);
 
 		auto rng = m_cache.equal_range(_key);
 		for (auto it = rng.first; it != rng.second; ++it)
 			if (_layout->isCompatibleUpToSet(COpenGLPipelineLayout::DESCRIPTOR_SET_COUNT-1u, it->second.layout.get())==(COpenGLPipelineLayout::DESCRIPTOR_SET_COUNT-1u))
 				return it->second.binary;
 		return {0,nullptr};
 	}
-	//assumes that m_cache does not already contain an item with key==_key and layout fully compatible with _val.layout
-	void insert(SCacheKey&& _key, SCacheVal&& _val, const asset::ICPUBuffer* _spirv)
+	const spirv_cross::ParsedIR* findParsedSpirv(const std::array<uint64_t, 4>& _key)
```

Comment:
maybe make the argument an `ICPUShader` then the `ICPUShader` can keep track of its own hash?

Reply:
why? pipeline operates on specialized shaders so why even bother with unspecialized ones?

Reply:
but you're findind the parsed spirv of the unspecialized shader right?

Reply:
theres no spirv of unspecialized shader and spirv of specialized shader. Theres just spirv. Specialized shader is spirv + spec info

Reply:
well we can compute the hash of spir-v and GLSL the same anyway, and make ICPUShader keep that.